src/passes/OnceReduction.cpp

```diff
@@ -0,0 +1,439 @@
+        optInfo.onceFuncs.at(curr->name) = global;
+
+        // We can ignore the get in the "once" pattern at the top of the
+        // function.
+        readGlobals[global]--;
+      }
+    }
+
+    for (auto& kv : readGlobals) {
+      auto global = kv.first;
+      auto count = kv.second;
+      if (count > 0) {
+        // This global has reads we cannot reason about, so do not optimize it.
+        optInfo.onceGlobals.at(global) = false;
+      }
+    }
+  }
+
+  // Check if a function body is in the "once" pattern. Return the name of the
+  // global if so, or an empty name otherwise.
+  //
+  // TODO: If the "once" function is inlined, this pattern can show up in random
+  //       places, and we can look for it there as well.
+  // TODO: Handle related patterns like if (!once) { .. }, but other opts will
+  //       tend to normalize to this form anyhow.
+  Name getOnceGlobal(Expression* body) {
+    // Look the pattern mentioned above:
+    //
+    //  function foo() {
+    //    if (foo$once) return;
+    //    foo$once = 1;
+    //    ...
+    //
+    auto* block = body->dynCast<Block>();
+    if (!block) {
+      return Name();
+    }
+    auto& list = block->list;
+    if (list.size() < 2) {
+      return Name();
+    }
+    auto* iff = list[0]->dynCast<If>();
+    if (!iff) {
+      return Name();
+    }
+    auto* get = iff->condition->dynCast<GlobalGet>();
+    if (!get) {
+      return Name();
+    }
+    if (!iff->ifTrue->is<Return>() || iff->ifFalse) {
+      return Name();
+    }
+    auto* set = list[1]->dynCast<GlobalSet>();
+
+    // Note that we have already checked the set's value earlier, but we do need
+    // it to not be unreachable (so it is actually set).
+    if (!set || set->name != get->name || set->type == Type::unreachable) {
+      return Name();
+    }
+    return get->name;
+  }
+
+private:
+  OptInfo& optInfo;
+};
+
+// Information in a basic block. We track relevant expressions, which are calls
+// calls to "once" functions, and writes to "once" globals.
+struct BlockInfo {
+  std::vector<Expression*> exprs;
+};
+
+// Performs optimization in all functions. This reads onceGlobalsSetInFuncs in
+// order to know what "once" globals are written by each function (so that when
+// we see a call, we can infer things), and when it is finishes with a function
+// it has learned which "once" globals it must set, and it then writes out
+// newOnceGlobalsSetInFuncs with that result. Later iterations will then use
+// those values in place of onceGlobalsSetInFuncs, which propagates things to
+// callers. This in effect mixes local optimization with the global
+// propagation - as we need to run the full local optimization in order to infer
+// the new values for onceGlobalsSetInFuncs, that is unavoidable (in principle,
+// we could also do a full propagation to a fixed point in between running
+// local optimization, but that would require more code - it might be more
+// efficient, though).
```

Comment:
```suggestion
// we see a call, we can infer things), and when it finishes with a function it
// has learned which "once" globals it must set, and it then writes out
// newOnceGlobalsSetInFuncs with that result. Later iterations will then use
// those values in place of onceGlobalsSetInFuncs, which propagate things to
// callers. This in effect mixes local optimization with the global propagation
// - as we need to run the full local optimization in order to infer the new
// values for onceGlobalsSetInFuncs, that is unavoidable (in principle, we could
// also do a full propagation to a fixed point in between running local
// optimization, but that would require more code - it might be more efficient,
// though).
```
(I only deleted the stray `is` and the rest is 80-col wrapping)

> as we need to run the full local optimization in order to infer the new values for onceGlobalsSetInFuncs, that is unavoidable

The 'optimization' is the part we convert operations with nops, and I don't think we have to run that to infer the new values..? (I've seen the comments and I understand you want to mix them; here I'm mostly confused about the comments)

> it might be more efficient, though.

Why?

Reply:
In addition to converting global sets and calls into nops, we also note which globals are set (by global sets and calls). That information is what we propagate to blocks dominated by us, but also to functions that call us.

It might be more efficient to propagate fully before doing more local work because the local work will go through all the basic blocks in each function, it's not just doing an operation on the graph of functions.

Reply:
> It might be more efficient to propagate fully before doing more local work because the local work will go through all the basic blocks in each function, it's not just doing an operation on the graph of functions.

When @tlively suggested to split those two, you said it's good to keep the status quo because this is more efficient.. (I think you said it's free) And the comment here says otherwise.

Reply:
Ah, good point, I was maybe inconsistent there. Looks like the original context was a response to @tlively saying this:

> It might be faster to first propagate the onceGlobalsSetInFuncs through the call graph (which doesn't require scanning through the blocks or expressions of a function) and only then start modifying functions.

Doing a first propagation through the call graph wouldn't speed things up, I don't think, as I said above, because to get the full info to propagate we have to scan into functions. We can learn new things from both operations, so we must cycle on both.

Separating a call graph propagation out *might* make sense. It would be more efficient if there is a deep call graph path that we need to get to - we'd get there immediately, instead of each step going through the function contents. I suspect that might be more common than the opposite case, where it would be slower - where we must do full function-level work at each step. That's why I wrote it might be more efficient to add more code for call graph propagation.

As it's not clear which is faster (or if it matters), I think the way with less code is best, as is implemented atm.